lib/spack/spack/cmd/uninstall.py

```diff
@@ -158,19 +159,22 @@ def installed_dependents(specs, env):
 
     env_hashes = set(env.all_hashes()) if env else set()
 
-    all_specs_in_db = spack.store.db.query()
+    # Ensure we stop traversal at input specs.
+    visited = set(s.dag_hash() for s in specs)
```

Comment:
@scheibelp it would be simpler to implement this with BFS, since then it's simply `traverse_nodes(specs, root=False, order="breadth")` instead of doing DFS on the children. The only issue is that with BFS we can't know the associated input spec in constant time. So, for backporting purposes this ugliness suffices, but to be honest, I seriously doubt anybody is really interested in this very detailed error message about how [matching spec i] needs [dependent spec j] to be uninstalled. So, in the other PR I opted for just collecting all dependents in a flat list instead of keeping track of what spec they are a dependent of.